### PR TITLE
Add MongoDB test application

### PR DESCRIPTION
**Pull request by AlexLLKong**
image

Test application to demonstrate rudimentary CRUD operations on locally hosted node application and node instance.

There isn't much error handling.

### CODE REVIEW

**Comment by nok-ko** [comment]
I've included `yargs` as a dependency for command-line argument parsing, and made the app automatically try to connect to the MongoDB instance at `mongodb://localhost:27017`, using the database called `test`. If it fails, it tells you about it:

https://user-images.githubusercontent.com/13428215/166613182-d953eb89-977d-4c73-af14-efcf43e87541.mp4

**Comment by nok-ko** [comment]
@AlexLLKong I think it's best to only open the connection once, and close it at the end of the app's lifetime – seems like that's the [best practice](https://stackoverflow.com/questions/14495975/why-is-it-recommended-not-to-close-a-mongodb-connection-anywhere-in-node-js-code).

**Comment by nok-ko** [comment]
I've added some `npm` scripts:
- `npm start` to start the server, using your local MongoDB instance
- `npm prod` to do the same using the instance at `db.nokko.me`
- `npm test` to run tests – currently none are defined, but I would like to use [Jest](https://jestjs.io) for this purpose in the future.

image

**Comment by nok-ko** [comment]
Don't have proper tests yet, but everything works circa the latest commit:

image

## TBD:
- [ ] Add code that initializes the database with all the collections we need, if they don't exist yet
    - (Need schema from Alex)
- [ ] Make a common `cleanup()` function that closes any open file handles/other resources in case the process needs to exit mid-run. Not such a big concern, but would be nice to not corrupt anything when exiting "nicely."
- [ ] Make the existing Express routes respond with semantic error codes when failing – currently all responses are `200 OK` and status is indicated in the response body. :eyes:

**Comment by AlexLLKong** [comment]
image
image

Using Insomnia to test endpoints. Postman was giving me issues and I'm too lazy to use `curl`. 

Next up will be logout followed by signup.

**Comment by AlexLLKong** [comment]
image

Logout "works" but I wasn't able to work out a test for when it fails. Which probably means I'm not doing it right.

File: mongoDbTest.js
Comment (nok-ko) [inline mongoDbTest.js:89]: Yep, still need to figure out how to keep secrets properly. Hng.

File: mongoDbTest.js
Comment (nok-ko) [inline mongoDbTest.js:199]: Might be worth keeping a log file for this kind of thing.

File: mongoDbTest.js
Comment (nok-ko) [inline mongoDbTest.js:224]: What's the fail condition here, exactly? When would this code path be reached?

**Review by nok-ko — COMMENTED** [review]
Good stuff! Are the Insomnia tests only on your local machine? Can you make them available to the rest of the team through the repo somehow?

File: mongoDbTest.js
Comment (nok-ko) [inline mongoDbTest.js:235]: `400` means the server thinks it was sent a malformed request – this is not the case here.

> The server cannot or will not process the request due to something that is perceived to be a client error (e.g., malformed request syntax, invalid request message framing, or deceptive request routing).

Some StackOverflow answers indicate that `422` could be the right error code here. (Maybe; if the error is due to there not being a session to log out of.)

> The 422 (Unprocessable Entity) status code means the server understands the content type of the request entity (hence a 415(Unsupported Media Type) status code is inappropriate), and the syntax of the request entity is correct (thus a 400 (Bad Request) status code is inappropriate) but was unable to process the contained instructions.

That said, I'm not 100% certain when these error cases would be hit. So, hm.

File: mongoDbTest.js
Comment (AlexLLKong) [reply mongoDbTest.js:224]: image
image
image


Currently the only case would be finding multiple user documents with the same email.

**Review by nok-ko — CHANGES_REQUESTED** [review]
Not sure what error code this should be, not sure when the error case would be hit… not sure of much of anything, but this shouldn't be a `400`. Minor quibble, really – don't worry.

File: mongoDbTest.js
Comment (nok-ko) [reply mongoDbTest.js:224]: I see! Worth documenting in the code, and sending back a more detailed error message if possible.

**Comment by AlexLLKong** [comment]
image
image

Making a user account works. No sanitization at the moment though

File: mongoDbTest.js
Comment (nok-ko) [inline mongoDbTest.js:252]: Shouldn't the kit be created without any involvement from the code calling into the API?

File: public/js/mongoTestClient.js
Comment (nok-ko) [inline public/js/mongoTestClient.js:20]: I guess whether we send the request data in the body or the query parameters doesn't really matter, but I thought we were going to use query parameters. 😅 

If you don't want to change it, just say so.

Comment (nok-ko) [reply public/js/mongoTestClient.js:20]: Oh, actually, don't change this! It makes sense for this to be in a body parameter, at least from reading [this blog post](https://wahlnetwork.com/2017/09/25/working-with-restful-api-query-body-and-path-parameters/).

General rule seems to be:
- Path parameters for getting a specific thing out of the API
    - For example: **GET** `/people/katy_p`
- Query parameters for filtering, or otherwise limiting the output of a query for lots of things
    - For example: **GET** `/people/katy_p/friends?limit=10&nameLike=Alex`
- Body parameters for supplying data needed to fulfill a request to the endpoint
    - For example: **POST** `/create-cat`, body: `{name:"Fluffy"}`

File: mongoDbTest.js
Comment (AlexLLKong) [reply mongoDbTest.js:252]: Is a user able to start building a kit before signing up? If not, then I guess that can be changed to a default skeleton object

Comment (nok-ko) [reply mongoDbTest.js:252]: What's the expected value for the `kit` property in a user's document? I think this might just be me being confused, I should take a look at the schema.

**Comment by nok-ko** [comment]
Once #9 is merged, we should rebase this branch on top of `main` and make sure everything still works.

File: mongoDbTest.js
Comment (nok-ko) [reply mongoDbTest.js:89]: Going to write a proper secret-grabbing thing after #9 is merged.

File: mongoDbTest.js
Comment (nok-ko) [inline mongoDbTest.js:151]: Is `$set` a convention in the MongoDB library, or a property name you came up with yourself? Should probably read more about the Node Mongo lib… 😅

File: mongoDbTest.js
Comment (nok-ko) [inline mongoDbTest.js:159]: Status codes?

File: mongoDbTest.js
Comment (nok-ko) [inline mongoDbTest.js:182]: Status codes…

File: mongoDbTest.js
Comment (nok-ko) [inline mongoDbTest.js:196]: Better message would be “Could not find user.” – also, actually we should be sending internal names like `userNotFound` and translating them on the frontend.

File: mongoDbTest.js
Comment (nok-ko) [inline mongoDbTest.js:210]: Would be nice to explicitly document when the session saving might fail – I don't know if it's important to handle the error case here or not.

Comment (nok-ko) [reply mongoDbTest.js:210]: Actually – if the failure means the user doesn't get logged in, then sending “Logged in.” is lying to the frontend!

File: mongoDbTest.js
Comment (nok-ko) [reply mongoDbTest.js:235]:
```suggestion
				res.status(422).send("Unable to logout");
```

File: mongoDbTest.js
Comment (nok-ko) [reply mongoDbTest.js:182]:
```suggestion
		res.status(/*XXX*/).send({ status: "fail", msg: errMsg ? errMsg : "could not delete person." });
```

File: mongoDbTest.js
Comment (nok-ko) [reply mongoDbTest.js:159]:
```suggestion
		res.status(/*XXX*/).send({ status: "fail", msg: "could not change name." });
```

File: mongoDbTest.js
Comment (nok-ko) [inline mongoDbTest.js:244]: Should make an issue for this, do it later.

File: mongoDbTest.js
Comment (nok-ko) [inline mongoDbTest.js:265]: Resolve this `TODO` if it's a quick change/doesn't require further work/won't be clobbered by future work, otherwise just make an issue for this and do the redirect later.

File: mongoDbTest.js
Comment (nok-ko) [inline mongoDbTest.js:280]: Small improvement, we can do this so that devs can Ctrl/Cmd-click on a URL to go right to the test server :)

```suggestion
  console.log(`Server listening on http://localhost:${port} !`);
```

File: package.json
Comment (nok-ko) [inline package.json:9]: Should remove Jay's old MySQL code, maybe in this PR if it's quick or in another one if not.

**Review by nok-ko — CHANGES_REQUESTED** [review]
Mostly good, take a look at my suggestions though.

File: mongoDbTest.js
Comment (AlexLLKong) [reply mongoDbTest.js:151]: It just tells MongoDB to set the field `name` in this instance to the value of `newName`

https://www.mongodb.com/docs/manual/reference/operator/update/set/

File: mongoDbTest.js
Comment (AlexLLKong) [reply mongoDbTest.js:125]: I don't actually know when the session could fail to save but I changed it so that at least we aren't lying if it fails.

File: mongoDbTest.js
Comment (AlexLLKong) [reply mongoDbTest.js:158]: #11

**Comment by nok-ko** [comment]
AT LAST! Latest commit should super-duper fix #8. @AlexLLKong if you could make up a random session secret and use my new `shakeguardSecrets.js` module to load it in, that'd be great.

Here:
```js
app.use(session(
	{
	  secret: "shhhh...secret", // load it in from the secrets object instead!
	  name: "ShakeGuardSessionID",
	  resave: false,
	  // create a unique identifier for that client
	  saveUninitialized: true
	})
  );
```

**Review by nok-ko — APPROVED** [review]
Right, that seems to fix everything I brought up, excellent work. Maybe do a nice thing for our front-end folks and document what all the messages mean somewhere – `API.md` or something? Maybe there's a better way, or maybe it's a bit early to put effort into that kind of thing, but would be nice. 👍

File: package.json
Comment (nok-ko) [inline package.json:13]: Shoot, forgot:

Comment (nok-ko) [reply package.json:13]:
```suggestion
    "prestart": "node downloadSecrets.mjs",
    "preprod": "node downloadSecrets.mjs",
```

**Review by nok-ko — COMMENTED** [review]
Whoops.

**Pull request merged**
